sotodlib/io/bookbinder.py

```diff
@@ -563,6 +481,91 @@ def replace_times_and_trim_frame(self, f):
+        if self._prev_smurf_sample is None:
+            prev_frame_end = self.BOOK_START_TIME.time - 1
+        else:
+            prev_frame_end = self._prev_smurf_sample.time
+        # Define a strict upper limit (exclusive) up until which to look/interpolate for missing samples
+        if self.BOOK_END_TIME is not None:
+            end_time = min(flush_time.time, self.BOOK_END_TIME.time)
+        elif self._smurf_timestamps is not None:
+            end_time = min(flush_time.time, self._smurf_timestamps[-1] + 1)
+        else:
+            end_time = min(flush_time.time, data.times[-1].time + 1)
+
+        # Obtain list of filled-in timestamps, whether from provided list or estimate with fill_time_gaps
+        if self._smurf_timestamps is not None:
+            # trim the list of timestamps to the current frame
+            m_ = np.logical_and(self._smurf_timestamps < end_time, self._smurf_timestamps > prev_frame_end)
+            ts = self._smurf_timestamps[m_]
+        else:
+            # fallback when no timestamp is provided. We switch to estimate sample interval based on data
+            assert data.times[0].time >= prev_frame_end
+            # Use fill_time_gaps to fill in any missing samples in the middle, as well as
+            # at the beginning (between previous frame and current frame)
+            dt = np.median(np.diff(data.times))  # estimate sample interval
+            if self._prev_smurf_sample is None:
+                # Use estimated sample interval to fill backward to prev_frame_end = book start time - 1
+                # since book start time does not usually line up with sample interval
+                ts = np.append(np.flip(np.arange(data.times[0].time-dt, prev_frame_end, -dt, dtype=int)), data.times)
+                ts = fill_time_gaps(ts)
+            elif (data.times[0].time - prev_frame_end)/dt > 0.5:
+                ts = fill_time_gaps(np.insert(data.times, 0, prev_frame_end))[1:]
+            else:
+                ts = fill_time_gaps(np.asarray(data.times))
+            # Use estimated sample interval to fill until end_time. Not using fill_time_gaps because
+            # end_time is externally determined and does not usually line up with sample interval
+            ts = np.append(ts, np.arange(data.times[-1].time+dt, end_time, dt, dtype=int))
+
+        # Create new G3SuperTimestream with filled-in samples
+        if len(data.times) < len(ts):
+            dt = np.median(np.diff(ts))
+            i_missing, _ = find_missing_samples(ts, np.array(data.times), atol=dt/2)
+            m = np.ones(len(ts), dtype=bool)
+            m[i_missing] = False
+            assert np.sum(m) == len(data.times)
+            new_data = np.full((data.data.shape[0], len(ts)), self.FLAGGED_SAMPLE_VALUE, dtype=data.data.dtype)
+            new_data[:,m] = data.data
+            dataout = so3g.G3SuperTimestream(data.names, core.G3VectorTime(ts), new_data)
+            flag_gap = core.G3VectorBool(~m)[:self.maxlength]
```

Comment:
I'm confused why this `self.maxlength` limit is here but isn't applied to the data array or the flag_gap array on line 562. Is this needed? It seems like by construction flag_gap and new_data should already be the same length.

Reply:
It’s there because the frame length may exceed maxlength after filling in missing samples, and while the data array gets passed through `split_timestream` after returning, the flag array is not. _However_, I think this is superseded by

https://github.com/simonsobs/sotodlib/blob/a4f4c68bcca22c4623ff449963c2d7a6a7bf03b2/sotodlib/io/bookbinder.py#L610

and I simply forgot to remove it (since it doesn’t complain even when len(flag_gap) < self.maxlength).